File: Bit_Manipulation/LC_338.py
# Written by: Jian-Yuan Lin
# Date: 2022.12.06

class Solution:
    def countBits(self, n: int) -> List[int]:
        # O(n) solution using DP
        # even more understandable solution by Neetcode
        dp = [0] * (n+1)
        offset = 1
        for i in range(1, n+1):
            # advance to the new most significant bit
            if offset * 2 == i:
                offset = i
            dp[i] = 1 + dp[i-offset]
        return dp
        
        
        """
        O(nlogn) solution 
        
        def count(n):
            ans = 0
        
            while n:
                ans += 1
                n = n & (n-1) # eleminate the least significant 1 bit to 0!
            return ans
        ans = []
        for i in range(n+1):
            ans.append(count(i))
        return ans
        """
        

File: Bit_Manipulation/LC_338.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.12.06

class Solution {
public:
    int count_bit(int n){
        int ans = 0;
        while (n){
            ans++;
            n = n & (n-1);
        }
        return ans;
    }
    vector<int> countBits(int n) {
        // DP solution is very smart!
        vector<int> dp(n+1, 0);
        int offset = 1;
        for (int i = 1; i <= n; i++){
            // advanced to the next bit
            if (offset * 2 == i){
                offset = i;
            }
            dp[i] = 1 + dp[i-offset];
        }
        return dp;
        /* using LC 191 solution
        vector<int> ans;
        for(int i = 0; i <= n; i++){
            ans.push_back(count_bit(i));
        }
        return ans;*/
    }
};